topics/dapp-wallet-msg.md

```diff
@@ -0,0 +1,110 @@
+# Messages
+
+## Permission
+
+A `permission_request` message is sent in order to establish a connection:
+1. The relying party sends `permission_request` to the wallet. The message includes:
+    - basic information about the relying party (`appMetadata`),
+    - a list of networks on which the relying party plans to operate (`networks`),
+    - a list of permission scopes it requires (`scopes`).
+2. The wallet receives the message, presents the details of the to-be-established connection to the user and asks the user to select the account that will be paired in response.
+    - If the user approves the request, the wallet sends `permission_response` back to the relying party. The message includes:
+        - basic information about the wallet (`appMetadata`),
+        - a list of networks on which the user has agreed the relying part can operate, should be a subset of the equivalent from `permission_request` (`networks`),
+        - a list of permission scopes the user has agreed the relying party can be granted, should be a subset of the equivalent from `permission_request` (`scopes`)
+        - the account details (princial or public key (*TBD what is required*) and optional subaccount) of the user selected account (`account`)
+    - If the user rejects the request, the wallet sends `error` back to the relying party with appropriate explanation.
+
+```
+{ // Request
+    type: "permission_request"
+    appMetadata: /* dApp */ {
+        name: String
+        icon: Optional<String>
+    }
+    networks: Array<{
+        type: "mainnet" | "testnet"
+        name: Optional<String>
+        rpcUrl: Optional<String>
+    }>
+    scopes: Array<"canister_call">
+}
+
+{ // Response
+    type: "permission_response"
+    appMetadata: /* wallet */ {
+        name: String
+        icon: Optional<String>
+    }
+    networks: Array<{
+        type: "mainnet" | "testnet"
+        name: Optional<String>
+        rpcUrl: Optional<String>
+    }>
+    scopes: Array<"canister_call">
+    account: {
```

Comment:
Something like
```
{ // Response
    type: "permission_response"
    appMetadata: /* wallet */ { ... }
    networks: Array<{ ... }>
    scopes: Array<...>
    principal: String
    ledger: Optional<{
        subaccount: Optional<String>
    }>
}
```
?

Reply:
Yes, exactly. However, we need to be careful. Just because the wallet claims to have access to `principal` does not mean it actually does (hence why I proposed a challenge response scheme in the last working group session).

Alternatively, we could also reduce the set of information provided here (i.e. the wallet only gives feedback that the user approved the connection without giving information about the users identity). But this would make it very cumbersome for use-cases where the relying needs access to the principal in order to build the arguments for the request...

Reply:
> However, we need to be careful. Just because the wallet claims to have access to principal does not mean it actually does (hence why I proposed a challenge response scheme in the last working group session).

Ah, yes. Sorry, I forgot to include the challenge in my response, but I do remember about it 😄 

> Alternatively, we could also reduce the set of information provided here (i.e. the wallet only gives feedback that the user approved the connection without giving information about the users identity). But this would make it very cumbersome for use-cases where the relying needs access to the principal in order to build the arguments for the request...

From the UX point of view, users could want to know which identity they used to sign in to the relying parties, meaning that, most probably, the relying parties would always request to share the identity anyway. So unless there are some privacy-preserving concerns, I would keep the principal in the response. To keep it simple 😄